File: front_dev/src/domain/admin/troubleshooting.ts
// トラブルシューティング記事管理の型定義（ADR-0079 / Knowledge MCP の
// TroubleshootingSummary / TroubleshootingDetail に対応）。新規作成・削除は初期スコープ外。

export interface TroubleshootingSummary {
  id: number;
  source_key: string;
  title: string;
  subtitle: string | null;
  tags: string[];
  source_updated_at: string | null;
}

export interface TroubleshootingTagRef {
  id: number;
  name: string;
}

export interface TroubleshootingDetail {
  id: number;
  source_key: string;
  title: string;
  subtitle: string | null;
  symptom: string | null;
  operation_history: string | null;
  error_code: string | null;
  error_message: string | null;
  system_environment: string | null;
  hardware_environment: string | null;
  version_info: string | null;
  guidance: string;
  cause: string | null;
  notes: string | null;
  keyword_raw: string | null;
  body_html: string;
  source_updated_at: string | null;
  created_at: string | null;
  updated_at: string | null;
  tags: TroubleshootingTagRef[];
}

export interface TroubleshootingListResponse {
  items: TroubleshootingSummary[];
  total: number;
}

export interface TroubleshootingListParams {
  keyword?: string;
  source_key?: string;
  tag_id?: number[];
  limit?: number;
  offset?: number;
}

// 未指定フィールドは「変更なし」。tag_ids は [] で全解除、未指定で変更なし。
// source_key / body_html / source_updated_at は編集対象外。
export interface TroubleshootingUpdateRequest {
  title?: string;
  subtitle?: string | null;
  symptom?: string | null;
  operation_history?: string | null;
  error_code?: string | null;
  error_message?: string | null;
  system_environment?: string | null;
  hardware_environment?: string | null;
  version_info?: string | null;
  guidance?: string;
  cause?: string | null;
  notes?: string | null;
  keyword_raw?: string | null;
  tag_ids?: number[];
}

export interface SourceKeyListResponse {
  source_keys: string[];
}

File: knowledge_mcp/src/knowledge_mcp/models/troubleshooting.py
"""トラブルシューティング記事の管理用データモデル（ADR-0079 / REQ-202609071337 9章）。

TroubleshootingSummary（一覧行）/ TroubleshootingDetail（詳細編集）と、業務エラー
TroubleshootingError を定義する。QA 管理（models/qa.py）と同様の構成。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from typing import List, Optional


class TroubleshootingError(Exception):
    """トラブルシューティング記事編集の業務エラー（対象不存在、存在しない tag_ids 指定等）。"""


def _iso(value: Optional[datetime]) -> Optional[str]:
    return value.isoformat() if value is not None else None


@dataclass
class TroubleshootingSummary:
    id: int
    source_key: str
    title: str
    subtitle: Optional[str] = None
    tags: List[str] = field(default_factory=list)
    source_updated_at: Optional[datetime] = None

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "source_key": self.source_key,
            "title": self.title,
            "subtitle": self.subtitle,
            "tags": list(self.tags),
            "source_updated_at": _iso(self.source_updated_at),
        }


@dataclass
class TroubleshootingDetail:
    id: int
    source_key: str
    title: str
    guidance: str
    subtitle: Optional[str] = None
    symptom: Optional[str] = None
    operation_history: Optional[str] = None
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    system_environment: Optional[str] = None
    hardware_environment: Optional[str] = None
    version_info: Optional[str] = None
    cause: Optional[str] = None
    notes: Optional[str] = None
    keyword_raw: Optional[str] = None
    body_html: str = ""
    source_updated_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    tags: List[dict] = field(default_factory=list)  # [{"id": int, "name": str}, ...]

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "source_key": self.source_key,
            "title": self.title,
            "subtitle": self.subtitle,
            "symptom": self.symptom,
            "operation_history": self.operation_history,
            "error_code": self.error_code,
            "error_message": self.error_message,
            "system_environment": self.system_environment,
            "hardware_environment": self.hardware_environment,
            "version_info": self.version_info,
            "guidance": self.guidance,
            "cause": self.cause,
            "notes": self.notes,
            "keyword_raw": self.keyword_raw,
            "body_html": self.body_html,
            "source_updated_at": _iso(self.source_updated_at),
            "created_at": _iso(self.created_at),
            "updated_at": _iso(self.updated_at),
            "tags": list(self.tags),
        }
